File: 0503 next-greater-element-ii/solution.java
class Solution {
    public int[] nextGreaterElements(int[] nums) {
        int n=nums.length;
        int[] res = new int[n];
        int[] st= new int[n];
        Arrays.fill(res,-1);
        int top=-1;

        for(int i=0;i<2*nums.length;i++) {
            int num = nums[i%n];
            while(top!=-1 && nums[st[top]]<num) {
                res[st[top--]]=num;
            }

            if(i<n)
             st[++top]=i; 
        }
        return res;
    }
}

File: 0503 next-greater-element-ii/solution.go
func nextGreaterElements(nums []int) []int {
	top, n := -1, len(nums)
	res := make([]int, n)
	st := make([]int, n)
	for i, _ := range res {
		res[i] = -1
	}
	for i := 0; i < 2*n; i++ {
		num := nums[i%n]
		for top != -1 && nums[st[top]] < num {
			res[st[top]] = num
			top--
		}
		if i < n {
			top++
			st[top] = i
		}
	}
	return res
}
